save-demo/src/main/kotlin/com/saveourtool/save/demo/service/GithubDownloadToolService.kt

```diff
@@ -0,0 +1,132 @@
+
+    private fun getGithubMetadataUrlByKey(key: ToolKey) = with(key) {
+        val release = if (vcsTagName == LATEST_VERSION) {
+            vcsTagName
+        } else {
+            "tags/$vcsTagName"
+        }
+        "$GITHUB_API_URL/$ownerName/$toolName/releases/$release"
+    }
+
+    private fun getMetadata(key: ToolKey): ReleaseMetadata {
+        val channel: Channel<ReleaseMetadata> = Channel()
+        scope.launch {
+            httpClient.get(getGithubMetadataUrlByKey(key))
+                .bodyAsText()
+                .let {
+                    jsonSerializer.decodeFromString<ReleaseMetadata>(it)
+                }
+                .let {
+                    channel.send(it)
+                    channel.close()
+                }
+        }
+        return runBlocking { channel.receive() }
+    }
+
+    @Suppress("ReactiveStreamsUnusedPublisher")
+    private suspend fun downloadAsset(asset: ReleaseAsset): Flux<ByteBuffer> = httpClient.get {
+        url(asset.downloadUrl)
+        accept(asset.contentType())
+    }
+        .bodyAsChannel()
+        .toByteBufferFlux()
+
+    private fun downloadFromGithubAndUploadToStorage(key: ToolKey) = getMetadata(key).assets
+        .filterNot(ReleaseAsset::isDigest)
+        .first()
+        .let { asset ->
+            scope.launch {
+                toolStorage.upload(key, downloadAsset(asset)).subscribe()
+            }
+                .invokeOnCompletion { exception ->
+                    exception?.let {
+                        if (it is CancellationException) {
+                            logger.debug("Download of ${key.toPrettyString()} was cancelled.")
+                        } else {
+                            logger.error("Error while downloading ${key.toPrettyString()}: ${it.message}")
+                        }
+                    } ?: logger.debug("${key.toPrettyString()} was successfully downloaded.")
+                }
+        }
+
+    @PostConstruct
+    private fun loadToStorage() = toolStorage.list()
+        .collectList()
+        .flatMapIterable { availableFiles ->
+            supportedTools.filter {
+                it !in availableFiles
+            }
+                .also { tools ->
+                    if (tools.isEmpty()) {
+                        logger.debug("All required tools are already present in storage.")
+                    } else {
+                        val toolsToBeDownloaded = tools.joinToString(", ") { it.toPrettyString() }
+                        logger.info("Tools to be downloaded: [$toolsToBeDownloaded]")
+                    }
+                }
+        }
+        .flatMap { key ->
+            downloadFromGithubAndUploadToStorage(key).toMono()
+        }
+        .subscribe()
+
+    companion object {
+        @Suppress("GENERIC_VARIABLE_WRONG_DECLARATION")
+        private val logger = getLogger<GithubDownloadToolService>()
+        private const val GITHUB_API_URL = "https://api.github.com/repos"
```

Comment:
why github? maybe download artifact from maven central?

Reply:
AFAIU, maven central is mostly used by java developers while GitHub is popular among the majority of developers

Reply:
> why github? maybe download artifact from maven central?

Because neither we nor _KtLint_ maintainers publish executable _fat jars_ to _Maven Central_.

This is about downloading release assets in terms of GitHub, not about transitive dependencies in Gradle/Maven/IOvy sense.

Reply:
> AFAIU, maven central is mostly used by java developers while GitHub is popular among the majority of developers regardless the language

agree, good point

Reply:
> ktlint uploads https://repo1.maven.org/maven2/com/pinterest/ktlint/0.47.1/ktlint-0.47.1-all.jar to maven central.

Thanks, was unaware of that.

So, they're using the `all` qualifier.
Probably we should do the same with _Diktat_ fat JAR, too.